addon/hooks/pagination.ts

```diff
@@ -0,0 +1,283 @@
+
+        //set configs from initial args
+        delete args.context;
+        delete args.modelName;
+        delete args.rows;
+        delete args.metadata;
+        delete args.sorts;
+        this.setConfigs(args);
+    }
+
+    /**
+     * Sets various pagination configurations
+     * @param {PaginationConfigs} args
+     */
+    @action
+    setConfigs(config: PaginationConfigs) {
+        this.config = { ...this.config,  ...config };
+        this.hasMore = this.rows.length >= this.config.limit!;
+    }
+
+    /**
+     * Utility method for completely replacing the current rows array/metadata
+     * @param {NativeArray<T> | T[]} rows
+     * @param {M} metadata
+     */
+    @action
+    setRows(rows: NativeArray<T> | T[], metadata?: M) {
+        this.rows = rows;
+        this.metadata = metadata;
+    }
+
+    /**
+     * Builds the query params object and makes the request for the next
+     * page of results (or the first page, if reset is true)
+     * @param {Boolean} reset
+     * @returns {Promise<T[]>}
+     */
+    @action
+    async loadModels(reset = false): Promise<T[]> {
+        if(reset) {
+            this.clearModels();
+        }
+
+        const queryParams = buildQueryParams({
+            context: this.context,
+            offset: this.offset,
+            sorts: this.sorts,
+            limit: this.config.limit,
+            filterList: this.config.filterList,
+            includeList: this.config.includeList,
+            pagingRootKey: this.config.pagingRootKey,
+            filterRootKey: this.config.filterRootKey,
+            includeKey: this.config.includeKey,
+            sortKey: this.config.sortKey,
+            serverDateFormat: this.config.serverDateFormat,
+            processQueryParams: this.config.processQueryParams
+        });
+
+        try {
+            this.isLoading = true;
+            const result = await this.queryModels(queryParams);
+            const rows = result.toArray();
+            this.hasMore = rows.length >= this.config.limit!;
+            this.metadata = result.meta;
+            this.rows.pushObjects(rows);
+            return rows;
+        } finally {
+            this.isLoading = false;
+        }
+    }
+
+    /**
+     * Makes the store.query() request using the provided query params object
+     * @param {any} queryParams
+     * @returns {Promise<RecordArrayWithMeta<T>>}
+     */
+    @action
+    async queryModels(queryParams: any): Promise<RecordArrayWithMeta<T>> {
+        const results = await this.store.query(this.modelName, queryParams) as RecordArrayWithMeta<T>;
+        return results;
+    }
+
+    /**
+     * Loads the next page of models if there are more to load and is not currently loading
+     * @returns {Promise<T[]> | null}
+     */
+    @action
+    loadMoreModels(): Promise<T[]> | null {
+        if (this.hasMore && !this.isLoadingModels) {
+            return this.loadModels();
+        }
+
+        return null;
```

Comment:
@billdami is there a reason we are returning `null` here instead of undefined (and just having the `if`)

Reply:
its because of TS's "Not all code paths return a value." error you get otherwise, even if you explicitly define the return type of this method here as `Promise<T[]> | undefined `, which I think is a bit weird.

Do you know of a better way to suppress those warnings?

Reply:
Are we positive its TS or is it es-lint? Either way, I wonder if it would work if you just did `return undefined` at the bottom

Reply:
Its TS, this is the full warning text:
> Not all code paths return a value. ts(7030)

But yup, just switching it to `return undefined` works.